File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = max(nums)
        curr_max , curr_min = 1,1
        for num in nums:
            temp = curr_max *num
            curr_max = max(num*curr_max,  num*curr_min , num)
            curr_min = min(temp ,num*curr_min , num)
            res = max(res,curr_max)
        return res
        

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int curr_max = nums[0];
        int curr_min = nums[0];
        int res = nums[0];
        for (int i = 1; i < n; i++) {
            int temp = curr_max;
            curr_max = max({nums[i], nums[i] * curr_max, nums[i] * curr_min});
            curr_min = min({nums[i], nums[i] * temp, nums[i] * curr_min});
            res = max(res, curr_max);
        }
        return res;
    }
};
